lib/os/hash_map_oa_lp.c

```diff
@@ -0,0 +1,326 @@
+#include <zephyr/sys/util.h>
+
+enum bucket_state {
+	UNUSED,
+	USED,
+	TOMBSTONE,
+};
+
+struct oalp_entry {
+	uint64_t key;
+	uint64_t value;
+	enum bucket_state state;
+};
+
+static struct oalp_entry *sys_hashmap_oa_lp_find(const struct sys_hashmap *map, uint64_t key,
+						 bool used_ok, bool unused_ok, bool tombstone_ok)
+{
+	size_t i;
+	size_t j;
+	struct oalp_entry *entry = NULL;
+	const size_t n_buckets = map->data->n_buckets;
+	uint32_t hash = map->hash_func(&key, sizeof(key));
+	struct oalp_entry *const buckets = map->data->buckets;
+
+	if (buckets == NULL) {
+		return NULL;
+	}
+
+	__ASSERT_NO_MSG(n_buckets != 0);
+
+	for (i = 0, j = hash % n_buckets; i < n_buckets; ++i, ++j, j %= n_buckets) {
+		__ASSERT_NO_MSG(j < n_buckets);
+
+		entry = &buckets[j];
+
+		switch (entry->state) {
+		case USED:
+			if (used_ok && entry->key == key) {
+				return entry;
+			}
+			break;
+		case UNUSED:
+			if (unused_ok) {
+				return entry;
+			}
+			break;
+		case TOMBSTONE:
+			if (tombstone_ok) {
+				return entry;
+			}
+			break;
+		default:
+			__ASSERT(false, "Invalid entry state. Memory has been corrupted");
+			break;
+		}
+	}
+
+	return NULL;
+}
+
+static int sys_hashmap_oa_lp_insert_no_rehash(struct sys_hashmap *map, uint64_t key, uint64_t value,
+					      uint64_t *old_value)
+{
+	int ret;
+	struct oalp_entry *entry = NULL;
+	struct sys_hashmap_oa_lp_data *data = (struct sys_hashmap_oa_lp_data *)map->data;
+
+	entry = sys_hashmap_oa_lp_find(map, key, true, true, true);
+	__ASSERT_NO_MSG(entry != NULL);
+
+	switch (entry->state) {
+	case UNUSED:
+		++data->size;
+		ret = 1;
+		break;
+	case TOMBSTONE:
+		--data->n_tombstones;
+		++data->size;
+		ret = 0;
+		break;
+	case USED:
+	default:
+		ret = 0;
+		break;
+	}
+
+	if (old_value != NULL) {
+		*old_value = entry->value;
+	}
+
+	entry->state = USED;
+	entry->key = key;
+	entry->value = value;
+
+	return ret;
+}
+
+/*
+ * This is similar to sys_hashmap_should_rehash(), but here we also take the
+ * number of TOMBSTONE entries into account in the 'growing' comparison.
+ */
+static bool should_rehash(const struct sys_hashmap *map, bool grow, size_t *new_num_buckets)
+{
+	bool shrink = !grow;
+	struct sys_hashmap_oa_lp_data *data = (struct sys_hashmap_oa_lp_data *)map->data;
+	size_t new_size = grow ? data->size + 1 : data->size;
+
+	/* maximum size imposed by the implementation */
+	__ASSERT_NO_MSG(new_size < SIZE_MAX / 100);
+
+	if (shrink && new_size == 0) {
+		/* last entry is being removed. Implementations should free storage */
+		*new_num_buckets = 0;
+		return true;
+	} else if (grow && data->n_buckets == 0) {
+		/* grow to a suitable initial size */
+		*new_num_buckets = ceiling_fraction(new_size * 100, map->config->load_factor);
+		return true;
+	} else if (grow && ((new_size + data->n_tombstones) * 100) / data->n_buckets >=
+				   map->config->load_factor) {
+		/* grow at a rate of 2x */
+		*new_num_buckets = data->n_buckets << 1;
+		return true;
```

Comment:
Aha! Suggestion: grow by 1.5x, not 2x.

https://github.com/facebook/folly/blob/main/folly/docs/FBVector.md#memory-handling

Reply:
Are you thinking about rehashing in place? That seemed a bit more challenging tbh.

Reply:
Ohh...oh right. Yeah nevermind my comment. I was thinking "Oh I wonder what growth factor we used" and forgot that this isn't a grow-in-place situation. I guess I'm not even sure how rehashing in-place would work, would definitely be challenging.